### PR TITLE
54867 env var editable in radix console

### PR DESCRIPTION
**Pull request by satr**

### CODE REVIEW

File: src/api/resource-environment-variable.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export const makeUrl = (appName, envName, componentName, envVarName) => {
+  const encAppName = encodeURIComponent(appName);
+  const encEnvName = encodeURIComponent(envName);
+  const encComponentName = encodeURIComponent(componentName);
+  const encEnvVarName = encodeURIComponent(envVarName);
+  return `/applications/${encAppName}/environments/${encEnvName}/components/${encComponentName}/envvars/${envVarName}`;
```
Comment (nilsgstrabo) [inline src/api/resource-environment-variable.js:6]: Use encoded **encEnvVarName** in path, not **envVarName**

Comment (nilsgstrabo) [reply src/api/resource-environment-variable.js:6]: Is this file used at all? I can't find any imports.

File: src/components/async-resource/index.js [hunk 2]
```diff
@@ -17,7 +17,12 @@ const AsyncResource = ({
   loading,
   resource,
   resourceParams,
+  disableSync,
```
Comment (nilsgstrabo) [inline src/components/async-resource/index.js:20]: Can't see that disableSync is used anywhere

File: src/components/environment-variables/environment-variables-list.js [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    }
+    setInEditMode(false);
+    setContext({ paused: false });
+  };
+  const handleReset = () => {
+    resetState();
+    setInEditMode(false);
+    setContext({ paused: false });
+  };
+  function getOriginalEnvVarToolTip(envVar) {
```
Comment (nilsgstrabo) [inline src/components/environment-variables/environment-variables-list.js:78]: Can we move all these functions outside of the **EnvironmentVariablesList** function, e.g. to line 17.
They make it a bit hard to follow the flow in **EnvironmentVariablesList**

File: src/components/environment-variables/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+const EnvironmentVariables = (props) => {
+  const {
+    appName,
+    envName,
+    componentName,
+    componentType,
+    includeRadixVars,
+    readonly,
+  } = props;
+  const [context, setContext] = useState({ paused: false });
```
Comment (nilsgstrabo) [inline src/components/environment-variables/index.js:14]: React has a concept named context (https://reactjs.org/docs/context.html) which has a different usage. Perhaps context and useContext should be renamed to it's actual usage; enabling and disabling envvar polling=> const [pollingEnabled, setPollingEnabled] and a bool instead of an object.

File: src/components/environment-variables/use-poll-env-vars.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { usePollingJson } from '../../effects';
+import envVarsNormaliser from '../../models/environment-variable/normaliser';
+
+const usePollEnvVars = (appName, envName, componentName, context) => {
```
Comment (nilsgstrabo) [inline src/components/environment-variables/use-poll-env-vars.js:4]: Should it be a bool parameter for context, and a more descriptive name, instead of an object? Ref https://github.com/equinor/radix-web-console/pull/356/files#r693699731

File: src/components/environment-variables/environment-variables-list.js [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+                                    return [...editableEnvVars];
+                                  })
+                                }
+                              />
+                            </div>
+                          </Table.Cell>
+                        </Table.Row>
+                      );
+                    }
+                    if (includeRadixVars === true) {
```
Comment (nilsgstrabo) [inline src/components/environment-variables/environment-variables-list.js:260]: `if (includeRadixVars) {` is sufficient

File: src/components/environment-variables/index.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  const [pollEnvVarsState] = usePollEnvVars(
+    appName,
+    envName,
+    componentName,
+    context
+  );
+  return (
+    <EnvironmentVariablesList
```
Comment (nilsgstrabo) [inline src/components/environment-variables/index.js:22]: Handle loading state and errors for pollEnvVarsState. If it fails, no error is displayed

File: src/components/environment-variables/style.css [hunk 8]
```diff
@@ -0,0 +1,28 @@
+.form-field {
+  display: flex;
+  flex-wrap: wrap;
+  width: 100%;
+
+  --form-field--bool-size: 1rem;
```
Comment (nilsgstrabo) [inline src/components/environment-variables/style.css:6]: Is this needed?

File: src/models/environment-variable/editable-environment-variable.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+import PropTypes from 'prop-types';
+import EnvVarModel from '.';
+
+export default Object.freeze({
```
Comment (nilsgstrabo) [inline src/models/environment-variable/editable-environment-variable.js:4]: Seems like this isnot used anywhere. Can it be removed?

File: src/api/resource-environment-variables.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+export const makeUrl = (appName, envName, componentName) => {
```
Comment (nilsgstrabo) [inline src/api/resource-environment-variables.js:1]: The exports in this file is only imported and used through redux/sagas, but to me it doesn't look like envvar state is used from redux. envvars are read using usePollEnvVars.

File: src/api/resources.js [hunk 11]
```diff
@@ -18,6 +19,7 @@ export const apiResources = {
   DEPLOYMENT: deployment,
   DEPLOYMENTS: deployments,
   ENVIRONMENT: environments,
+  ENVVARS: environmentVariables,
```
Comment (nilsgstrabo) [inline src/api/resources.js:22]: Can be removed if redux state for envvars are not used.

File: src/api/resource-environment-variable.js (hunk 1 above)
Comment (satr) [reply src/api/resource-environment-variable.js:6]: No, it is outdated (replaced with `resource-environment-variables.js`). Deleted now.

File: src/components/environment-variables/environment-variables-list.js [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+              </>
+            )}
+          </div>
+        </form>
+      )}
+    </React.Fragment>
+  );
+};
+
+export default EnvironmentVariablesList;
```
Comment (nilsgstrabo) [inline src/components/environment-variables/environment-variables-list.js:287]: Would be useful to define `.propTypes` for the parameters of EnvironmentVariablesList

File: src/models/environment-variable/updatable-environment-variable.js [hunk 13]
```diff
@@ -0,0 +1,6 @@
+import PropTypes from 'prop-types';
+
+export default Object.freeze({
```
Comment (nilsgstrabo) [inline src/models/environment-variable/updatable-environment-variable.js:3]: Only imported by use-save-env-var.js, but the import is not used there. Can probably be deleted.

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
- Unused code should be removed.
- handle async loading state; show error, loading spinner
and a few minor issues

File: src/components/async-resource/index.js (hunk 2 above)
Comment (satr) [reply src/components/async-resource/index.js:20]: It does not seem so anymore. Removed.

File: src/components/environment-variables/environment-variables-list.js (hunk 3 above)
Comment (satr) [reply src/components/environment-variables/environment-variables-list.js:78]: Moved all functions at the bottom of the class. Left only const-s.

File: src/components/environment-variables/index.js (hunk 4 above)
Comment (satr) [reply src/components/environment-variables/index.js:14]: renamed to `poolingState` and `setPoolingState`

File: src/components/environment-variables/use-poll-env-vars.js (hunk 5 above)
Comment (satr) [reply src/components/environment-variables/use-poll-env-vars.js:4]: Renamed to `poolingState`

File: src/components/environment-variables/environment-variables-list.js (hunk 6 above)
Comment (satr) [reply src/components/environment-variables/environment-variables-list.js:260]: I think this is more precise:

File: src/components/environment-variables/style.css (hunk 8 above)
Comment (satr) [reply src/components/environment-variables/style.css:6]: No. Removed.

File: src/models/environment-variable/editable-environment-variable.js (hunk 9 above)
Comment (satr) [reply src/models/environment-variable/editable-environment-variable.js:4]: Not anymore. Deleted.

File: src/api/resource-environment-variables.js (hunk 10 above)
Comment (satr) [reply src/api/resource-environment-variables.js:1]: File has been deleted now.

File: src/api/resources.js (hunk 11 above)
Comment (satr) [reply src/api/resources.js:22]: Deleted.

File: src/components/environment-variables/environment-variables-list.js (hunk 12 above)
Comment (satr) [reply src/components/environment-variables/environment-variables-list.js:287]: Added definitions of params

File: src/models/environment-variable/updatable-environment-variable.js (hunk 13 above)
Comment (satr) [reply src/models/environment-variable/updatable-environment-variable.js:3]: It is used in property definition
```
import updatableEnvVar from '../../models/environment-variable/updatable-environment-variable';
...
useSaveEnvVar.propTypes = {
...
  updatableEnvVars: PropTypes.arrayOf(PropTypes.shape(updatableEnvVar))
    .isRequired,
```

File: src/components/environment-variables/environment-variables-list.js (hunk 6 above)
Comment (nilsgstrabo) [reply src/components/environment-variables/environment-variables-list.js:230]: Thats fine

File: src/components/environment-variables/use-save-env-var.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+import { usePatchJson } from '../../effects';
+import PropTypes from 'prop-types';
+import updatableEnvVar from '../../models/environment-variable/updatable-environment-variable';
+
+const useSaveEnvVar = ({
+  appName,
+  envName,
+  componentName,
+  updatableEnvVars,
```
Comment (nilsgstrabo) [inline src/components/environment-variables/use-save-env-var.js:9]: updatableEnvVars is unused

File: src/components/environment-variables/use-save-env-var.js [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+}) => {
+  const encAppName = encodeURIComponent(appName);
+  const encEnvName = encodeURIComponent(envName);
+  const encComponentName = encodeURIComponent(componentName);
+  const path = `/applications/${encAppName}/environments/${encEnvName}/components/${encComponentName}/envvars`;
+
+  return usePatchJson(path, (props) => {
+    return props.updatableEnvVars;
```
Comment (nilsgstrabo) [inline src/components/environment-variables/use-save-env-var.js:17]: ref above comment: props.updatableEnvVars refers to the object you pass as argument to saveFunc in environment-variable-list.js. Since the request body of path is onløy the list of variables, you only need to pass the updatableEnvVars value in saveFunc and return it in the function in usePatchJson

File: src/components/environment-variables/environment-variables-list.js [hunk 16]
```diff
@@ -0,0 +1,307 @@
...
+    setPoolingState({ paused: true });
+    setInEditMode(true);
+  };
+  const handleSave = () => {
+    if (readonly) {
+      return;
+    }
+    const updatableEnvVars = getUpdatableEnvVars(editableEnvVars);
+    if (updatableEnvVars.length > 0) {
+      saveFunc({ appName, envName, componentName, updatableEnvVars });
```
Comment (nilsgstrabo) [inline src/components/environment-variables/environment-variables-list.js:61]: You only need to pass updatableEnvVars to saveFunc, and use it directly in usePatchJson. See comment in use-save-env-var.js

**Review by nilsgstrabo — CHANGES_REQUESTED** [review]
A few comments related to useSaveEnvVar

File: src/components/environment-variables/use-save-env-var.js (hunk 14 above)
Comment (satr) [reply src/components/environment-variables/use-save-env-var.js:9]: I was not sure it should not exist as variable here. Removed.

File: src/components/environment-variables/environment-variables-list.js (hunk 16 above)
Comment (satr) [reply src/components/environment-variables/environment-variables-list.js]: Fixed

File: src/components/environment-variables/use-save-env-var.js (hunk 15 above)
Comment (satr) [reply src/components/environment-variables/use-save-env-var.js]: Fixed

**Review by nilsgstrabo — APPROVED** [review]
